pkg/ottl/ottlfuncs/func_replace_all_patterns_test.go

```diff
@@ -59,8 +59,68 @@ func Test_replaceAllPatterns(t *testing.T) {
 			},
 			function: optionalArg,
 			want: func(expectedMap pcommon.Map) {
-				expectedMap.PutStr("test", "4804d6b7f03268e33f78c484977f3d81771220df07cc6aac4ad4868102141fad world")
-				expectedMap.PutStr("test2", "4804d6b7f03268e33f78c484977f3d81771220df07cc6aac4ad4868102141fad")
+				expectedMap.PutStr("test", "hash(hello {universe}) world")
+				expectedMap.PutStr("test2", "hash(hash(hello {universe}))") // hash applied twice because of replace_all_patterns
+				expectedMap.PutStr("test3", "goodbye world1 and world2")
```

Comment:
Could we have a replacement function case case with pattern `(world\d)` and replacement `$1` (or similar)? I don't think we have anything to cover the case where we have two distinct matches from the same capture group.

Reply:
I think we will need to use `FindAllStringSubmatchIndex` and handle each individual submatch separately.

Reply:
this is a viable approach but doesn't seem to work. This is the [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/658c1794f37762cf9e48ef49167c9de4af678ded). It isn't clear how we would replace all match groups using after resolving and hashing them using `ReplaceAllString`. More complex than it seems .

Reply:
I think I figured it out, the latest [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/e1972e13f86e075ed2cf1318f8d72dd7cb8e2c8b) now supports hashing distinct matches from the same capture group using `strings.ReplaceAll`

Reply:
Only hash the capture group specified in the replacement string [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/ab286f15b16ff8734a551e824beb9d971d987b07). In the previous commit we were hashing all the capture groups.

Reply:
Resolved [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/7819437f67b6c511fb3a2f46bac74e485e6e9391)